### PR TITLE
feat(lib): flag to disable dictionary sort

### LINKED ISSUE #525: Stop sorting the fields in alphabetical order in generated yaml file
### Description of the feature or enhancement:

Hi, I don't know whether this feature already exist or not, Can someone please help me with this issue.

When I run `cdk8s synth` yaml files are getting generated in dist folder. List or directory declared in .ts file is getting sorted alphabetically in generated yaml file, is there a way to not sort these fields..?

### Use Case: Avoiding alphabetical sort of dictionary or list resolves some pods to be not stuck in **Creating Container**
```
let tenant = new Tenant(this, 'tenant', {
            metadata: {
                name: 'minio',
                creationTimestamp: null,
                labels: {
                    app: 'minio'
                },
                scheduler: {
                    name: ""
                }
            },
            spec: {
                console: {
                    consoleSecret: {
                        name: 'minio-console-secret'
                    },
                    image: 'minio/console:v0.4.6',
                    replicas: 2,
                    resources: {},
                },
                credsSecret: {
                    name: 'minio-creds-secret'
                },
                image: 'minio/minio:RELEASE.2020-12-23T02-24-12Z',
                mountPath: '/export',
                requestAutoCert: true,
                zones: [
                    {
                        resources: {},
                        servers: 4,
                        volumeClaimTemplate: {
                            apiVersion: 'v1',
                            kind: 'persistentvolumeclaims',
                            spec: {
                                storageClassName: 'provisioner',
                                accessModes: ['ReadWriteOnce'],
                                resources: {
                                    requests: {
                                        storage: '1G'
                                    }
                                }
                            }
                        },
                        volumesPerServer: 1,
                        affinity: {
                            podAntiAffinity: {
                                preferredDuringSchedulingIgnoredDuringExecution: [
                                    {
                                        weight: 1,
                                        podAffinityTerm: {
                                            labelSelector: {
                                                matchExpressions: [
                                                    {
                                                        key: 'app',
                                                        operator: 'In',
                                                        values: ['minio']
                                                    }
                                                ]
                                            },
                                            topologyKey: 'kubernetes.io/hostname'
                                        }
                                    }
                                ]
                            }
                        }
                    }]
            }
        });
```

The generated yaml file from above .ts file is 
```apiVersion: "minio.min.io/v1"
kind: "Tenant"
metadata:
  labels:
    app: "minio"
  name: "minio"
  namespace: "minio"
  scheduler:
    name: ""
spec:
  console:
    consoleSecret:
      name: "minio-console-secret"
    image: "minio/console:v0.4.6"
    replicas: 2
    resources: {}
  credsSecret:
    name: "minio-creds-secret"
  image: "minio/minio:RELEASE.2020-12-23T02-24-12Z"
  mountPath: "/export"
  requestAutoCert: true
  zones:
    - affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - podAffinityTerm:
                labelSelector:
                  matchExpressions:
                    - key: "app"
                      operator: "In"
                      values:
                        - "minio"
                topologyKey: "kubernetes.io/hostname"
              weight: 1
      resources: {}
      servers: 4
      volumeClaimTemplate:
        apiVersion: "v1"
        kind: "persistentvolumeclaims"
        spec:
          accessModes:
            - "ReadWriteOnce"
          resources:
            requests:
              storage: "1G"
          storageClassName: "provisioner"
      volumesPerServer: 
```
      
### Proposed Solution: Generate YAML file without alphabetical sort

```apiVersion: "minio.min.io/v1"
kind: "Tenant"
metadata:
  labels:
    app: "minio"
  name: "minio"
  namespace: "minio"
  scheduler:
    name: ""
spec:
  console:
    consoleSecret:
      name: "minio-console-secret"
    image: "minio/console:v0.4.6"
    replicas: 2
    resources: {}
  credsSecret:
    name: "minio-creds-secret"
  image: "minio/minio:RELEASE.2020-12-23T02-24-12Z"
  mountPath: "/export"
  requestAutoCert: true
  zones:
    - resources: {}
      servers: 4
      volumeClaimTemplate:
        apiVersion: "v1"
        kind: "persistentvolumeclaims"
        spec:
          accessModes:
            - "ReadWriteOnce"
          resources:
            requests:
              storage: "1G"
          storageClassName: "provisioner"
      volumesPerServer: 1
      affinity:
      podAntiAffinity:
        preferredDuringSchedulingIgnoredDuringExecution:
        - weight: 1
          podAffinityTerm:
            labelSelector:
              matchExpressions:
                - key: "app"
                  operator: In
                  values:
                  - minio
            topologyKey: "kubernetes.io/hostname"

### PR DESCRIPTION
**Pull request by eladb**
Introduce an environment flag `CDK8S_DISABLE_SORT` which disables the sorting of keys in synthesized output. This is intentionally supported through an environment variable and not as a first-class API since relying on key ordering in dictionaries is considered an anti-pattern. See issue for discussion.

Resolves #525

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/cdk8s/package.json [hunk 1]
```diff
@@ -32,11 +32,10 @@
   "devDependencies": {
     "@types/follow-redirects": "^1.13.0",
     "@types/jest": "^26.0.20",
-    "@types/json-stable-stringify": "^1.0.32",
-    "@types/node": "^10.17.52",
+    "@types/node": "^10.17.0",
     "@types/yaml": "^1.9.6",
-    "@typescript-eslint/eslint-plugin": "^4.15.0",
```
Comment (iliapolo) [inline packages/cdk8s/package.json:38]: Why the downgrades here?

File: packages/cdk8s/src/api-object.ts [hunk 2]
```diff
@@ -166,17 +165,31 @@ export class ApiObject extends Construct {
 
   /**
    * Renders the object to Kubernetes JSON.
+   *
+   * To disable sorting of dictionary keys in output object set the
+   * `CDK8S_DISABLE_SORT` environment variable.
```
Comment (iliapolo) [inline packages/cdk8s/src/api-object.ts:170]:
```suggestion
   * `CDK8S_DISABLE_SORT` environment variable to any non-empty value.
```

File: packages/cdk8s/src/api-object.ts [hunk 3]
```diff
@@ -166,17 +165,31 @@ export class ApiObject extends Construct {
...
+    const sortKeys = process.env.CDK8S_DISABLE_SORT ? false : true;
+    const json = sanitizeValue(resolve(data), { sortKeys });
+    const patched = JsonPatch.apply(json, ...this.patches);
+
+    // reorder top-level keys so that we first have "apiVersion", "kind" and
+    // "metadata" and then all the rest
+    const result: any = {};
+    const orderedKeys = ['apiVersion', 'kind', 'metadata', ...Object.keys(patched)];
+    for (const k of orderedKeys) {
+      if (k in patched) {
```
Comment (iliapolo) [inline packages/cdk8s/src/api-object.ts:187]: When can this be `false`?

File: packages/cdk8s/package.json (hunk 1 above)
Comment (eladb) [reply packages/cdk8s/package.json:38]: That's the projen synth output. We need to fix in projen if we want to keep this file untampered

File: packages/cdk8s/src/api-object.ts (hunk 3 above)
Comment (eladb) [reply packages/cdk8s/src/api-object.ts:187]: Probably never, but just in case one of the primary fields was not set

Comment (iliapolo) [reply packages/cdk8s/src/api-object.ts:187]: Wouldn't it be better to throw in that case? Seems like an illegal situation.

Comment (eladb) [reply packages/cdk8s/src/api-object.ts:187]: Why illegal? There could be an object without metadata for example.

File: packages/cdk8s/package.json (hunk 1 above)
Comment (iliapolo) [reply packages/cdk8s/package.json:38]: Ok, can you just disable them from the auto upgrade job for now?

File: packages/cdk8s/src/api-object.ts (hunk 3 above)
Comment (iliapolo) [reply packages/cdk8s/src/api-object.ts:187]: Hmm. Can there? I thought these 3 are mandatory for all objects according to k8s spec.

Comment (eladb) [reply packages/cdk8s/src/api-object.ts:187]: I think metadata is technically optional (and even apiVersion for things like `List` if I recall)

Comment (iliapolo) [reply packages/cdk8s/src/api-object.ts:187]: Keep forgetting about those. ok.

**Review by iliapolo — APPROVED** [review]
Added `do-not-merge` for the upgrade workflow changes, but we can also get around to it some other time.

**Pull request merged**